repository/repository-core/src/main/java/org/eclipse/vorto/repository/core/impl/ModelRepositoryFactory.java

```diff
@@ -273,23 +297,50 @@ public IModelRepository getRepositoryByModel(ModelId modelId, IUserContext userC
     };
   }
 
+  /**
+   * This method accesses a request-scoped {@link UserNamespaceRolesCache} bean that maps composite
+   * workspace+user IDs to roles. <br/>
+   * This implies values for multiple calls with same workspace ID and username are cached within
+   * the current request, and expire in the next one. <br/>
+   * Not a perfect system as there are also numerous individual ID calls per request at times,
+   * but better than nothing (i.e. at least caches the large amount of repeated calls within e.g. a
+   * request to load a model).
+   *
+   * @param workspaceId
+   * @param username
+   * @return
+   */
   private Set<IRole> getUserRoles(String workspaceId, String username) {
-    // TODO improve caching for non anon users
+
     if (UserContext.isAnonymous(username)) {
       return new HashSet<>();
     }
 
-    String namespace = namespaceService.findNamespaceByWorkspaceId(workspaceId).getName();
-    try {
-      Set<IRole> userRoles = new HashSet<>(userNamespaceRoleService.getRoles(username, namespace));
-      if (userRepositoryRoleService.isSysadmin(username)) {
-        userRoles.add(RepositoryRole.SYS_ADMIN);
-      }
-      return userRoles;
-    } catch (DoesNotExistException e) {
-      LOGGER.debug("User or namespace not found. ", e);
-      return Collections.emptySet();
-    }
+    String compositeID = workspaceId.concat(username);
+
+    Set<IRole> result = userNamespaceRolesCache.get(compositeID)
```

Comment:
I would probably place this logic inside the UserNamespaceRoleService, so it would be automatically used, wherever namespace roles are required.

Reply:
> I would probably place this logic inside the UserNamespaceRoleService, so it would be automatically used, wherever namespace roles are required.

It does make sense to broaden it, but given the context (request-scoped caching) I suspect there might not be many more scenarios availing of the cache.

Reply:
Fair enough - I couldn't tell if there are many more scenarios where this is helpful - we can also keep it as is and measure the impact first

Reply:
The results for the specific test looked very good! I still think we should consider to implement a proper repository caching since this would improve the performance in general (given that there is a reasonable eviction strategy).

Reply:
> Fair enough - I couldn't tell if there are many more scenarios where this is helpful - we can also keep it as is and measure the impact first

I still think broadening per your suggestion is a good idea, if anything, for the "form" of it. Since this will have to wait a day or two, we have room to experiment.

Reply:
> The results for the specific test looked very good! I still think we should consider to implement a proper repository caching since this would improve the performance in general (given that there is a reasonable eviction strategy).

Sounds great. As we know, the caching may be tricky to implement given the situation but ultimately it is the Spring out-of-the-box tool for this matter, so it will likely be better than any homemade solution.